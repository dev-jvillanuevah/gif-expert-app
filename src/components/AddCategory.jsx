import { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState('');

  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const onSubmitValue = (event) => {
    event.preventDefault();
    const value = inputValue.trim();
    if (value.length < 1) return;

    onNewCategory(value);
    setInputValue('');
  };

  return (
    <form onSubmit={onSubmitValue} aria-label="form">
      <input
        type="text"
        placeholder="Search gifs"
        value={inputValue}
        onChange={onInputChange}
        //   onChange={event => onInputChange(event)} I send the first prop I can do it like the one on top
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
