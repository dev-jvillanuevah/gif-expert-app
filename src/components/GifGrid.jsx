import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifItem } from './GifItem';

export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {isLoading && <h2>Loading...</h2>}{' '}
      {/* If isLoading is true it will show the second argument (<h2>Loading...</h2>) */}
      <div className="card-grid">
        {' '}
        {/* This will now show until we have the images loaded */}
        {images.map((image) => (
          <GifItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};
