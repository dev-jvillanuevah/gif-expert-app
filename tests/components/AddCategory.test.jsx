import { fireEvent, render, screen } from '@testing-library/react';
import { AddCategory } from '../../src/components';

describe('tests on AddCagetory', () => {
  test('should match with the snapshot', () => {
    const { container } = render(<AddCategory onNewCategory={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  test('should change the value on the textbox', () => {
    render(<AddCategory onNewCategory={() => {}} />); // create the test subject
    const input = screen.getByRole('textbox'); // substract the input
    fireEvent.input(input, { target: { value: 'Goku' } }); // pass the input value firing the input event
    expect(input.value).toBe('Goku'); // make the assertion
  });

  // we did it this way instead of evaluate the onSubmitValue func
  // because we want to test the component
  test('should should call onNewCategory if the input has a value', () => {
    const inputValue = 'Goku';
    const onNewCategory = jest.fn();

    render(<AddCategory onNewCategory={onNewCategory} />);
    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');

    fireEvent.input(input, { target: { value: inputValue } });
    fireEvent.submit(form);

    expect(input.value).toBe('');

    expect(onNewCategory).toHaveBeenCalled();
    expect(onNewCategory).toHaveBeenCalledTimes(1);
    expect(onNewCategory).toHaveBeenCalledWith(inputValue);
  });

  test('should not call onNewCategory if the input has an empty value', () => {
    const onNewCategory = jest.fn();

    render(<AddCategory onNewCategory={onNewCategory} />);

    const form = screen.getByRole('form');

    fireEvent.submit(form);

    expect(onNewCategory).toBeCalledTimes(0);
    expect(onNewCategory).not.toHaveBeenCalled();
  });
});
