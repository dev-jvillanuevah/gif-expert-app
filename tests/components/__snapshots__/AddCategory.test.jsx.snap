// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on AddCagetory should match with the snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Search gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
